Fix intermittent failure in the "supports specifying a custom origin" viewer integration test

timvandermeij opened the pull request:
The problem in the original code is that `getTextAt` is called too early and therefore returns unexpected text content. This can happen because we call `increaseScale` and then wait for the page's text layer to be visible, but it can take some time before the zoom actually occurs/completes in the viewer and in the meantime the old (pre-zoom) text layer may still be visible, causing us to continue too soon because we don't validate that we're dealing with the post-zoom text layer.

This commit fixes the issue by simply waiting for the expected text to show up at the given origin coordinates, which makes the test work independent of viewer actions/timing.

Fixes #19951.

timvandermeij commented:
/botio integrationtest

Snuffleupagus commented on test/integration/viewer_spec.mjs:
```diff
@@ -81,10 +65,9 @@ describe("PDF viewer", () => {
-          expect(textAfterZoomIn)
-            .withContext(`In ${browserName}, zoom in`)
-            .toBe(text);
+          await page.waitForFunction(
+            `document.elementFromPoint(${originX}, ${originY})?.textContent === "${text}"`
+          );
```
> This can happen because we call `increaseScale` and then wait for the page's text layer to be visible, but it can take some time before the zoom actually occurs/completes in the viewer and in the meantime the old (pre-zoom) text layer may still be visible, causing us to continue too soon because we don't validate that we're dealing with the post-zoom text layer.

Possibly a stupid question, but isn't there a theoretical risk that these new checks could also target the "old" textLayer if the timing is just right?
Would it be even safer if we awaited the "textlayerrendered" event, before checking the textContent?

timvandermeij committed:
Fix intermittent failure in the "supports specifying a custom origin" viewer integration test

The problem in the original code is that `getTextAt` is called too
early and therefore returns unexpected text content. This can happen
because we call `increaseScale` and then wait for the page's text layer
to be visible, but it can take some time before the zoom actually
occurs/completes in the viewer and in the meantime the old (pre-zoom)
text layer may still be visible, causing us to continue too soon because
we don't validate that we're dealing with the post-zoom text layer.

This commit fixes the issue by simply waiting for the expected text to
show up at the given origin coordinates, which makes the test work
independent of viewer actions/timing.

timvandermeij commented:
/botio integrationtest

calixteman reviewed (approved):
LGTM. Thank you.

Snuffleupagus commented on test/integration/viewer_spec.mjs:
```diff
@@ -42,23 +42,41 @@ describe("PDF viewer", () => {
+        (scaleFactor, origin) => {
+          window.PDFViewerApplication.pdfViewer.updateScale({
+            drawingDelay: 0,
+            steps: 1,
```
This line shouldn't be necessary, since `steps` isn't used when a `scaleFactor` is provided; [please see this code](https://github.com/mozilla/pdf.js/blob/712d502561a29df3378dd7fb7cca7a22e4a48ea9/web/pdf_viewer.js#L2289-L2301).
```suggestion
```

Snuffleupagus reviewed (approved):
r=me, thank you.

timvandermeij commented:
/botio integrationtest
